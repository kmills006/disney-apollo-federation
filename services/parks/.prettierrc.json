{
  "printWidth": 80,
  "semi": true,
  "trailingComma": "all",
  "arrowParens": "always",
  "singleQuote": true,
  "tabWidth": 2
}
