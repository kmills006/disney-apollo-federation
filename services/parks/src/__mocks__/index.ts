import { IPark, ParkId } from '../models/Park';

export const MOCK_PARKS: IPark[] = [
  {
    id: ParkId.EPCOT,
    name: 'Park One',
    permalink: 'park-one',
  },
  {
    id: ParkId.MAGIC_KINGDOM,
    name: 'Park Two',
    permalink: 'park-two',
  },
];
