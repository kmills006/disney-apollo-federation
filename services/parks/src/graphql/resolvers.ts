import * as O from 'fp-ts/lib/Option';
import * as P from 'fp-ts/lib/pipeable';
import { ResolverFunc, ResolveType } from '@disney-federation/gql-utils';

import { IPark, ParkId } from '../models/Park';
import { IParkError } from '../models/ParkError';
import { IResolverContext } from './context';

type IGraphQLError = Pick<IParkError, 'message'>;
type NoParkFoundError = IGraphQLError;

type ParkPayload = IPark | NoParkFoundError;
type ParkPayloadTypeName = 'Park' | 'NoParkFoundError';

interface IQueryResolvers {
  parks: ResolverFunc<IPark[], IResolverContext>;
  park: ResolverFunc<ParkPayload, IResolverContext, { id: ParkId }>
}

interface IParkPayloadResolvers {
  __resolveType: ResolveType<ParkPayload, ParkPayloadTypeName>;
}

export interface IResolvers {
  [key: string]: {};
  Query: IQueryResolvers;
  ParkPayload: IParkPayloadResolvers;
}

export const findTypeNameInParent = <T extends string, P = {}>(
  parent: P,
  mapping: string[][],
): T | null => {
  const [, typename] = mapping.find(([property]) => property in parent) || [null, null];

  return typename as T;
};

export const resolvers: IResolvers = {
  ParkPayload: {
    __resolveType: (parent) => findTypeNameInParent<ParkPayloadTypeName>(parent, [
      ['name', 'Park'],
      ['message', 'NoParkFoundError'],
    ]),
  },
  Query: {
    parks: (_, __, ctx) => ctx.repositories.parks.getParks(),
    park: (_, args, ctx) => P.pipe(
      ctx.repositories.parks.getParkByPermalink(args.id),
      O.getOrElseW(() => ({
        message: `Park with id [${args.id}] not found.`,
      })),
    ),
  },
};
