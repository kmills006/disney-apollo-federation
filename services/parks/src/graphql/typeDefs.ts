import { gql } from 'apollo-server';

export const typeDefs = gql`
  type Query {
    parks: [Park!]!
    park(id: ParkId!): ParkPayload!
  }

  interface Error {
    message: String!
  }

  type NoParkFoundError implements Error {
    message: String!
  }

  union ParkPayload = Park | NoParkFoundError

  type Park @key(fields: "permalink") {
    id: ParkId!
    name: String!
    permalink: String!
  }

  enum ParkId {
    ANIMAL_KINGDOM
    EPCOT
    HOLLYWOOD_STUDIOS
    MAGIC_KINGDOM
  }
`;
