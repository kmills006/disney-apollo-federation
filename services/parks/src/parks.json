[
  {
    "name": "Magic Kingdom",
    "permalink": "magic-kingdom"
  },
  {
    "name": "Epcot",
    "permalink": "epcot"
  },
  {
    "name": "Animal Kingdom",
    "permalink": "animal-kingdom"
  },
  {
    "name": "Hollywood Studios",
    "permalink": "hollywood-studios"
  }
]
