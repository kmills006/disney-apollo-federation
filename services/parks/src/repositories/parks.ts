import * as A from 'fp-ts/lib/Array';
import * as O from 'fp-ts/lib/Option';
import * as P from 'fp-ts/lib/pipeable';

import { parseJsonFile } from '../db';
import {
  IPark,
  ParkId,
  ParkPermalink,
  RawPark,
} from '../models/Park';

const parkIdToPermalinkMap: Record<ParkId, ParkPermalink> = {
  [ParkId.ANIMAL_KINGDOM]: 'animal-kingdom',
  [ParkId.EPCOT]: 'epcot',
  [ParkId.HOLLYWOOD_STUDIOS]: 'hollywood-studios',
  [ParkId.MAGIC_KINGDOM]: 'magic-kingdom',
};

export interface IParkRepository {
  getParks: () => IPark[];
  getParkByPermalink: (id: ParkId) => O.Option<IPark>;
}

export const repository = (parks: IPark[]) => (): IParkRepository => ({
  getParks: () => parks,
  getParkByPermalink: (id) => (
    O.fromNullable(parks.find((p) => p.permalink === parkIdToPermalinkMap[id]))
  ),
});

/// // Move out of here
const addIdToParkRecord = (park: RawPark): O.Option<IPark> => {
  const id = (Object.keys(parkIdToPermalinkMap) as ParkId[])
    .find((key) => parkIdToPermalinkMap[key] === park.permalink as ParkPermalink);

  return !id ? O.none : O.some({ id, ...park });
};

const addParkIds = (parks: RawPark[]): O.Option<IPark[]> => P.pipe(
  parks.map(addIdToParkRecord),
  A.filter(O.isSome),
  A.array.sequence(O.option),
);
/// // Move out of here

const parks = P.pipe(
  parseJsonFile<RawPark[]>('src/db/parks.json'),
  O.fromEither,
  O.chain(addParkIds),
  O.fold(() => [], (p) => p),
);

export const parkRepository = repository(parks);
