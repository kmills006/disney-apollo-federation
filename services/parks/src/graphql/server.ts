import { buildFederatedSchema } from '@apollo/federation';
import { ApolloServer } from 'apollo-server';
import { GraphQLSchema } from 'graphql';

import { context, IResolverContext } from './context';
import { resolvers } from './resolvers';
import { typeDefs } from './typeDefs';

interface ApolloServerConfiguration<C> {
  schema: GraphQLSchema;
  context: () => C;
  engine?: boolean;
}

export const constructApolloServer = <C = IResolverContext>(
  config: ApolloServerConfiguration<C>,
): ApolloServer => (
    new ApolloServer({
      schema: config.schema,
      context: () => config.context(),
      engine: config.engine || false,
    })
  );

export const server = constructApolloServer({
  context,
  schema: buildFederatedSchema([{ resolvers, typeDefs }]),
});

export const startApolloServer = async (s: ApolloServer) => {
  s.listen(4000).then(({ url }) => {
    console.log(`ðŸš€ Server ready at ${url}`);
  });
};
