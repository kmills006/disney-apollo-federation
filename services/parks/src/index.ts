import { getOrElseW, map } from 'fp-ts/Either';
import { pipe } from 'fp-ts/function';

import { db } from './db';
import { Park } from './model';
import { parkRepository } from './repository';
import { initiateApolloServer } from './graphql/server';

const PORT = process.env.PORT || 4001;

const server = pipe(
  db<Park[]>('src/parks.json'),
  map(parkRepository),
  map((repository) => initiateApolloServer(repository)),
  getOrElseW((err) => {
    console.error(`Unable to initiate Apollo Server: ${err.message}`);
    process.exit(1);
  }),
);

server
  .listen(PORT)
  .then(() => console.log(`Parks GraphQL running on port ${PORT}`));
