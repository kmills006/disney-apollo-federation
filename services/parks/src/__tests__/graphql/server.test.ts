import { buildFederatedSchema } from '@apollo/federation';
import { constructApolloServer } from '@disney-federation/gql-utils';
import { ApolloServer, gql } from 'apollo-server';
import { createTestClient } from 'apollo-server-testing';
import * as O from 'fp-ts/lib/Option';
import { mockDeep } from 'jest-mock-extended';

import { MOCK_PARKS } from '../../__mocks__';
import { resolvers } from '../../graphql/resolvers';
import { typeDefs } from '../../graphql/typeDefs';
import { IParkRepository } from '../../repositories/parks';

describe.only('graphql', () => {
  describe('server', () => {
    const mockParksRepository = mockDeep<IParkRepository>();

    let server: ApolloServer;

    beforeEach(() => {
      server = constructApolloServer({
        schema: buildFederatedSchema({ resolvers, typeDefs }),
        context: () => ({
          repositories: { parks: mockParksRepository },
        }),
      });
    });

    describe('queries', () => {
      describe('parks', () => {
        const GET_PARKS = gql`query { parks { id name permalink } }`;

        it.only('returns a list of parks', async () => {
          mockParksRepository.getParks.mockReturnValueOnce(MOCK_PARKS);

          const { query } = createTestClient(server);
          const actual = await query({ query: GET_PARKS });

          expect(actual).toMatchSnapshot();
        });
      });

      describe('park', () => {
        const GET_PARK = gql`
          query Park($id: ParkId!) {
            park(id: $id) {
              ... on Park { id name permalink }
              ... on NoParkFoundError { message }
            }
          }
        `;

        const mockPark = MOCK_PARKS[0];

        it('returns a park by its id', async () => {
          mockParksRepository.getParkByPermalink.mockReturnValueOnce(O.some(mockPark));

          const { query } = createTestClient(server);
          const actual = await query({
            query: GET_PARK,
            variables: { id: 'EPCOT' },
          });

          expect(actual).toMatchSnapshot();
        });

        it('returns a NoParkFoundError', async () => {
          mockParksRepository.getParkByPermalink.mockReturnValueOnce(O.none);

          const { query } = createTestClient(server);
          const actual = await query({
            query: GET_PARK,
            variables: { id: 'MAGIC_KINGDOM' },
          });

          expect(actual).toMatchSnapshot();
        });
      });
    });
  });
});
