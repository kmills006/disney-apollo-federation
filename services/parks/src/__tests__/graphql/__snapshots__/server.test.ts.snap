// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`graphql server queries parks returns a list of parks 1`] = `
Object {
  "data": Object {
    "parks": Array [
      Object {
        "id": "EPCOT",
        "name": "Park One",
        "permalink": "park-one",
      },
      Object {
        "id": "MAGIC_KINGDOM",
        "name": "Park Two",
        "permalink": "park-two",
      },
    ],
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
